cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -0,0 +1,491 @@
+    deps.querier
+        .query::<Empty>(
+            &WormholeQuery::VerifySignature {
+                data: observations.clone(),
+                observer_set_index,
+                signature: signature.clone(),
+                block_time: env.block.height,
+            }
+            .into(),
+        )
+        .context("failed to verify signature")?;
+
+    let quorum = deps
+        .querier
+        .query::<u32>(
+            &WormholeQuery::CalculateQuorum {
+                observer_set_index,
+                block_time: env.block.height,
+            }
+            .into(),
+        )
+        .and_then(|q| {
+            usize::try_from(q).map_err(|_| StdError::ConversionOverflow {
+                source: ConversionOverflowError::new("u32", "usize", q.to_string()),
+            })
+        })
+        .context("failed to calculate quorum")?;
+
+    let observations: Vec<Observation> =
+        from_binary(&observations).context("failed to parse `Observations`")?;
+
+    let events = observations
+        .into_iter()
+        .map(|o| {
+            handle_observation(
+                deps.branch(),
+                o,
+                observer_set_index,
+                quorum,
+                signature.clone(),
+            )
+        })
+        .filter_map(Result::transpose)
+        .collect::<anyhow::Result<Vec<_>>>()
+        .context("failed to handle `Observation`")?;
+
+    Ok(Response::new()
+        .add_attribute("action", "submit_observations")
+        .add_attribute("owner", info.sender)
+        .add_events(events))
+}
+
+fn handle_observation(
+    mut deps: DepsMut<WormholeQuery>,
+    o: Observation,
+    observer_set_index: u32,
+    quorum: usize,
+    sig: wormhole_bindings::Signature,
+) -> anyhow::Result<Option<Event>> {
+    if accounting::has_transfer(deps.as_ref(), o.key.clone()) {
+        bail!("transfer for key \"{}\" already committed", o.key);
+    }
+
+    let key = PENDING_TRANSFERS.key(o.key.clone());
+    let mut pending = key
+        .may_load(deps.storage)
+        .map(Option::unwrap_or_default)
+        .context("failed to load `PendingTransfer`")?;
+    let data = match pending
+        .iter_mut()
+        .find(|d| d.observer_set_index() == observer_set_index && d.observation() == &o)
+    {
+        Some(d) => d,
+        None => {
+            pending.push(Data::new(o.clone(), observer_set_index));
+            let back = pending.len() - 1;
+            &mut pending[back]
+        }
+    };
+
+    if !data.add_signature(sig) {
+        bail!("cannot submit duplicate signatures for the same observation");
+    }
+
+    if data.signatures().len() < quorum {
+        // Still need more signatures so just save the pending transfer data and exit.
+        key.save(deps.storage, &pending)
+            .context("failed to save pending transfers")?;
+
+        return Ok(None);
+    }
+
+    let (msg, _) = serde_wormhole::from_slice_with_payload(&o.payload)
```

Comment:
similarly, I think it'd be a little more optimal to return this failure before saving signatures.

Reply:
It's basically a question of whether we want to parse the payload every time or only once we might reach quorum.  

The value type for `PENDING_TRANSFERS` is a `Vec<transfer::Data>` so we can continue to add arbitrarily many different observations for the same key until one of them reaches quorum.  So in the case where we fail to parse the payload, we can fix the issue offline and then resubmit new observations for the same key.  Once the new observation reaches quorum it will get committed and everything else will be deleted.

If we make the assumption that an observation that reaches quorum is highly unlikely to have an invalid payload, which doesn't seem *that* unreasonable, then we can optimize for that case and only parse it once at the cost of delaying the error case.  This seems like a worthwhile tradeoff to me but I don't feel that strongly about it so I'm happy to change it if you still prefer that we check the payload every time.